File: python/src/pinch/dispatcher.py
from __future__ import annotations

import asyncio
import contextlib
from collections.abc import Callable
from dataclasses import dataclass

from .jitter import JitteredDelay
from .notifier import Notifier
from .queue import PendingTask, TaskQueue
from .runner import ClaudeRunner, RunnerInvocation
from .semaphore import AsyncSemaphore, ProjectSemaphore
from .types import (
    BlockReason,
    Clock,
    FinishedEvent,
    HookMap,
    PacerStats,
    ResolvedConfig,
    StartedEvent,
    TaskResult,
)
from .window import WorkingWindow

UPSTREAM_PAUSE_MS = 60 * 60 * 1000
STARTUP_JITTER_MAX_MS = 5 * 60 * 1000
IDLE_POLL_MS = 10 * 60 * 1000
SHORT_POLL_MS = 10_000
FAST_FAIL_THRESHOLD_MS = 2_000


@dataclass(slots=True)
class DispatcherDeps:
    config: ResolvedConfig
    queue: TaskQueue[TaskResult]
    global_sem: AsyncSemaphore
    project_sem: ProjectSemaphore
    window: WorkingWindow
    jitter: JitteredDelay
    runner: ClaudeRunner
    clock: Clock
    hooks: HookMap


@dataclass(slots=True)
class _Admission:
    item: PendingTask[TaskResult]
    global_release: Callable[[], None]
    project_release: Callable[[], None]


class Dispatcher:
    def __init__(self, deps: DispatcherDeps):
        self._deps = deps
        self._notifier = Notifier()
        self._running = False
        self._draining = False
        self._shutting_down = False
        self._active = 0
        self._spawns_since_last_wave = 0
        self._cooldown_ends_mono = 0.0
        self._upstream_pause_ends_mono = 0.0
        self._next_spawn_ready_mono = 0.0
        self._loop_task: asyncio.Task[None] | None = None
        self._active_tasks: set[asyncio.Task[None]] = set()

    def _spawn_task(self, pick: _Admission) -> None:
        task = asyncio.create_task(self._run_task(pick))
        self._active_tasks.add(task)
        task.add_done_callback(self._active_tasks.discard)

    def start(self) -> None:
        if self._running:
            return
        self._running = True
        self._next_spawn_ready_mono = (
            self._deps.clock.monotonic_ms() + self._deps.jitter.spawn_delay_ms()
        )
        self._loop_task = asyncio.create_task(self._loop())

    def enqueue(self, item: PendingTask[TaskResult]) -> None:
        if self._shutting_down:
            if not item.future.done():
                item.future.set_exception(RuntimeError("pinch: pacer is shutting down"))
            return
        self._deps.queue.enqueue(item)
        hook = self._deps.hooks.on_enqueued
        if hook is not None:
            hook(
                {
                    "task_id": item.id,
                    "project_id": item.project_id,
                    "queue_depth": self._deps.queue.length,
                }
            )
        self._notifier.notify()

    def stats(self) -> PacerStats:
        return PacerStats(
            queue_depth=self._deps.queue.length,
            global_active=self._deps.global_sem.in_use,
            active_projects=tuple(self._deps.project_sem.active_project_ids()),
            spawns_since_last_wave=self._spawns_since_last_wave,
            in_wave_cooldown=self._deps.clock.monotonic_ms() < self._cooldown_ends_mono,
            window_open=self._deps.window.is_open_now(),
        )

    async def drain(self) -> None:
        self._draining = True
        self._notifier.notify()
        while self._deps.queue.length > 0 or self._active > 0:
            await self._notifier.wait()

    async def shutdown(self) -> None:
        self._shutting_down = True
        self._draining = True
        self._notifier.notify()
        await self.drain()
        self._running = False
        self._notifier.notify()
        if self._loop_task is not None:
            with contextlib.suppress(asyncio.CancelledError):
                await self._loop_task

    async def _loop(self) -> None:
        try:
            while self._running:
                if (
                    self._shutting_down
                    and self._deps.queue.length == 0
                    and self._active == 0
                ):
                    self._running = False
                    break

                if self._deps.queue.length == 0:
                    await self._wait_for_signal(IDLE_POLL_MS)
                    continue

                if not self._deps.window.is_open_now():
                    wait_ms = (
                        self._deps.window.ms_until_next_open()
                        + self._deps.jitter.startup_jitter_ms(STARTUP_JITTER_MAX_MS)
                    )
                    self._emit_blocked("window-closed", wait_ms)
                    await self._wait_for_signal(wait_ms)
                    continue

                now_mono = self._deps.clock.monotonic_ms()

                if now_mono < self._upstream_pause_ends_mono:
                    wait_ms = int(self._upstream_pause_ends_mono - now_mono)
                    self._emit_blocked("upstream", wait_ms)
                    await self._wait_for_signal(wait_ms)
                    continue

                if now_mono < self._cooldown_ends_mono:
                    wait_ms = int(self._cooldown_ends_mono - now_mono)
                    self._emit_blocked("wave-cooldown", wait_ms)
                    await self._wait_for_signal(wait_ms)
                    continue

                if now_mono < self._next_spawn_ready_mono:
                    wait_ms = int(self._next_spawn_ready_mono - now_mono)
                    await self._wait_for_signal(wait_ms)
                    continue

                pick = self._find_admissible()
                if pick is None:
                    self._emit_blocked(self._diagnose_block())
                    await self._wait_for_signal(SHORT_POLL_MS)
                    continue

                self._deps.queue.remove_by_id(pick.item.id)
                self._active += 1
                spawn_mono = self._deps.clock.monotonic_ms()
                self._next_spawn_ready_mono = spawn_mono + self._deps.jitter.spawn_delay_ms()
                self._spawns_since_last_wave += 1
                if self._spawns_since_last_wave >= self._deps.config.pacing.wave_every_n:
                    self._cooldown_ends_mono = (
                        self._deps.clock.monotonic_ms()
                        + self._deps.jitter.wave_cooldown_ms()
                    )
                    self._spawns_since_last_wave = 0
                self._spawn_task(pick)
        finally:
            self._running = False
            self._notifier.notify()

    async def _run_task(self, pick: _Admission) -> None:
        item = pick.item
        started_at = self._deps.clock.now()
        start_mono = self._deps.clock.monotonic_ms()
        waited_ms = max(int(start_mono - item.enqueued_at_mono_ms), 0)

        started_event: StartedEvent = {
            "task_id": item.id,
            "project_id": item.project_id,
            "waited_ms": waited_ms,
            "global_active": self._deps.global_sem.in_use,
            "project_active": self._deps.project_sem.in_use_for(item.project_id),
        }
        on_started = self._deps.hooks.on_started
        if on_started is not None:
            on_started(started_event)

        try:
            timeout_ms = item.task.timeout_ms or self._deps.config.runner.task_timeout_ms
            extra = tuple(item.task.args) if item.task.args else ()
            inv = RunnerInvocation(
                prompt=item.task.prompt,
                cwd=item.task.cwd,
                extra_args=extra,
                timeout_ms=float(timeout_ms),
            )
            out = await self._deps.runner.execute(inv)
            finished_at = self._deps.clock.now()
            duration_ms = max(int(self._deps.clock.monotonic_ms() - start_mono), 0)
            result = TaskResult(
                task_id=item.id,
                project_id=item.project_id,
                exit_code=out.exit_code,
                stdout=out.stdout,
                stderr=out.stderr,
                duration_ms=duration_ms,
                waited_ms=waited_ms,
                started_at=started_at,
                finished_at=finished_at,
            )
            if not item.future.done():
                item.future.set_result(result)
            finished_event: FinishedEvent = {
                "task_id": item.id,
                "project_id": item.project_id,
                "waited_ms": waited_ms,
                "global_active": started_event["global_active"],
                "project_active": started_event["project_active"],
                "exit_code": out.exit_code,
                "duration_ms": duration_ms,
            }
            on_finished = self._deps.hooks.on_finished
            if on_finished is not None:
                on_finished(finished_event)

            if out.exit_code != 0 and duration_ms < FAST_FAIL_THRESHOLD_MS:
                self._next_spawn_ready_mono = self._deps.clock.monotonic_ms()

            if out.upstream:
                self._upstream_pause_ends_mono = (
                    self._deps.clock.monotonic_ms() + UPSTREAM_PAUSE_MS
                )
        except Exception as err:
            if not item.future.done():
                item.future.set_exception(err)
        finally:
            pick.project_release()
            pick.global_release()
            self._active -= 1
            self._notifier.notify()

    def _find_admissible(self) -> _Admission | None:
        if self._deps.global_sem.available <= 0:
            return None
        for item in self._deps.queue.snapshot():
            global_release = self._deps.global_sem.try_acquire()
            if global_release is None:
                return None
            project_release = self._deps.project_sem.try_acquire(item.project_id)
            if project_release is not None:
                return _Admission(
                    item=item, global_release=global_release, project_release=project_release
                )
            global_release()
        return None

    def _diagnose_block(self) -> BlockReason:
        if self._deps.global_sem.available <= 0:
            return "global-limit"
        active = self._deps.project_sem.active_project_ids()
        if len(active) >= self._deps.config.limits.max_active_projects:
            return "active-projects"
        return "project-limit"

    def _emit_blocked(self, reason: BlockReason, ms_until_retry: int | None = None) -> None:
        hook = self._deps.hooks.on_blocked
        if hook is None:
            return
        payload = {"reason": reason}
        if ms_until_retry is not None:
            payload["ms_until_retry"] = ms_until_retry  # type: ignore[assignment]
        hook(payload)  # type: ignore[arg-type]

    async def _wait_for_signal(self, max_ms: float) -> None:
        if max_ms <= 0:
            return
        notifier_wait = self._notifier.wait()
        sleep_task = asyncio.create_task(self._deps.clock.sleep(max_ms))
        done, pending = await asyncio.wait(
            {notifier_wait, sleep_task}, return_when=asyncio.FIRST_COMPLETED
        )
        for p in pending:
            p.cancel()
        for d in done:
            if d.cancelled():
                continue
            with contextlib.suppress(asyncio.CancelledError):
                d.exception()

File: node/src/dispatcher.ts
import { JitteredDelay } from "./jitter.js";
import { Notifier } from "./notifier.js";
import { TaskQueue, type PendingTask } from "./queue.js";
import { ClaudeRunner } from "./runner.js";
import { ProjectSemaphore, Semaphore } from "./semaphore.js";
import type {
  BlockReason,
  Clock,
  FinishedEvent,
  HookMap,
  ResolvedConfig,
  StartedEvent,
  Task,
  TaskResult,
} from "./types.js";
import { WorkingWindow } from "./window.js";

const UPSTREAM_PAUSE_MS = 60 * 60 * 1000;
const STARTUP_JITTER_MAX_MS = 5 * 60 * 1000;
const IDLE_POLL_MS = 10 * 60 * 1000;
const SHORT_POLL_MS = 10_000;
const FAST_FAIL_THRESHOLD_MS = 2_000;

export interface DispatcherStats {
  readonly queueDepth: number;
  readonly globalActive: number;
  readonly activeProjects: readonly string[];
  readonly spawnsSinceLastWave: number;
  readonly inWaveCooldown: boolean;
  readonly windowOpen: boolean;
}

export interface DispatcherDeps {
  readonly config: ResolvedConfig;
  readonly queue: TaskQueue<TaskResult>;
  readonly globalSem: Semaphore;
  readonly projectSem: ProjectSemaphore;
  readonly window: WorkingWindow;
  readonly jitter: JitteredDelay;
  readonly runner: ClaudeRunner;
  readonly clock: Clock;
  readonly hooks: HookMap;
}

export class Dispatcher {
  private readonly notifier = new Notifier();
  private running = false;
  private draining = false;
  private shuttingDown = false;
  private active = 0;
  private spawnsSinceLastWave = 0;
  private cooldownEndsMono = 0;
  private upstreamPauseEndsMono = 0;
  private nextSpawnReadyMono = 0;
  private loopPromise: Promise<void> | null = null;

  constructor(private readonly deps: DispatcherDeps) {}

  start(): void {
    if (this.running) return;
    this.running = true;
    this.nextSpawnReadyMono =
      this.deps.clock.monotonicMs() + this.deps.jitter.spawnDelayMs();
    this.loopPromise = this.loop();
  }

  enqueue(item: PendingTask<TaskResult>): void {
    if (this.shuttingDown) {
      item.reject(new Error("pinch: pacer is shutting down"));
      return;
    }
    this.deps.queue.enqueue(item);
    this.deps.hooks.onEnqueued?.({
      taskId: item.id,
      projectId: item.projectId,
      queueDepth: this.deps.queue.length,
    });
    this.notifier.notify();
  }

  stats(): DispatcherStats {
    return {
      queueDepth: this.deps.queue.length,
      globalActive: this.deps.globalSem.inUse,
      activeProjects: this.deps.projectSem.activeProjectIds(),
      spawnsSinceLastWave: this.spawnsSinceLastWave,
      inWaveCooldown: this.deps.clock.monotonicMs() < this.cooldownEndsMono,
      windowOpen: this.deps.window.isOpenNow(),
    };
  }

  async drain(): Promise<void> {
    this.draining = true;
    this.notifier.notify();
    while (this.deps.queue.length > 0 || this.active > 0) {
      await this.notifier.wait();
    }
  }

  async shutdown(): Promise<void> {
    this.shuttingDown = true;
    this.draining = true;
    this.notifier.notify();
    await this.drain();
    this.running = false;
    this.notifier.notify();
    if (this.loopPromise) await this.loopPromise;
  }

  private async loop(): Promise<void> {
    try {
      while (this.running) {
        if (this.shuttingDown && this.deps.queue.length === 0 && this.active === 0) {
          this.running = false;
          break;
        }

        if (this.deps.queue.length === 0) {
          await this.waitForSignal(IDLE_POLL_MS);
          continue;
        }

        if (!this.deps.window.isOpenNow()) {
          const waitMs =
            this.deps.window.msUntilNextOpen() +
            this.deps.jitter.startupJitterMs(STARTUP_JITTER_MAX_MS);
          this.emitBlocked("window-closed", waitMs);
          await this.waitForSignal(waitMs);
          continue;
        }

        const nowMono = this.deps.clock.monotonicMs();

        if (nowMono < this.upstreamPauseEndsMono) {
          const waitMs = this.upstreamPauseEndsMono - nowMono;
          this.emitBlocked("upstream", waitMs);
          await this.waitForSignal(waitMs);
          continue;
        }

        if (nowMono < this.cooldownEndsMono) {
          const waitMs = this.cooldownEndsMono - nowMono;
          this.emitBlocked("wave-cooldown", waitMs);
          await this.waitForSignal(waitMs);
          continue;
        }

        if (nowMono < this.nextSpawnReadyMono) {
          const waitMs = this.nextSpawnReadyMono - nowMono;
          await this.waitForSignal(waitMs);
          continue;
        }

        const pick = this.findAdmissible();
        if (!pick) {
          this.emitBlocked(this.diagnoseBlock());
          await this.waitForSignal(SHORT_POLL_MS);
          continue;
        }

        this.removeFromQueue(pick.item.id);
        this.active++;
        const spawnMono = this.deps.clock.monotonicMs();
        this.nextSpawnReadyMono = spawnMono + this.deps.jitter.spawnDelayMs();
        this.spawnsSinceLastWave++;
        if (this.spawnsSinceLastWave >= this.deps.config.pacing.waveEveryN) {
          this.cooldownEndsMono =
            this.deps.clock.monotonicMs() + this.deps.jitter.waveCooldownMs();
          this.spawnsSinceLastWave = 0;
        }
        void this.runTask(pick);
      }
    } finally {
      this.running = false;
      this.notifier.notify();
    }
  }

  private async runTask(pick: Admission): Promise<void> {
    const { item, globalRelease, projectRelease } = pick;
    const startedAt = this.deps.clock.now();
    const startMono = this.deps.clock.monotonicMs();
    const waitedMs = Math.max(startMono - item.enqueuedAtMonoMs, 0);

    const startedEvent: StartedEvent = {
      taskId: item.id,
      projectId: item.projectId,
      waitedMs,
      globalActive: this.deps.globalSem.inUse,
      projectActive: this.deps.projectSem.inUseFor(item.projectId),
    };
    this.deps.hooks.onStarted?.(startedEvent);

    try {
      const timeoutMs = item.task.timeoutMs ?? this.deps.config.runner.taskTimeoutMs;
      const out = await this.deps.runner.execute({
        prompt: item.task.prompt,
        ...(item.task.cwd !== undefined ? { cwd: item.task.cwd } : {}),
        ...(item.task.args !== undefined ? { extraArgs: item.task.args } : {}),
        timeoutMs,
      });
      const finishedAt = this.deps.clock.now();
      const durationMs = Math.max(this.deps.clock.monotonicMs() - startMono, 0);
      const result: TaskResult = {
        taskId: item.id,
        projectId: item.projectId,
        exitCode: out.exitCode,
        stdout: out.stdout,
        stderr: out.stderr,
        durationMs,
        waitedMs,
        startedAt,
        finishedAt,
      };
      item.resolve(result);
      const finishedEvent: FinishedEvent = {
        ...startedEvent,
        exitCode: out.exitCode,
        durationMs,
      };
      this.deps.hooks.onFinished?.(finishedEvent);

      if (out.exitCode !== 0 && durationMs < FAST_FAIL_THRESHOLD_MS) {
        this.nextSpawnReadyMono = this.deps.clock.monotonicMs();
      }

      if (out.upstream) {
        this.upstreamPauseEndsMono = this.deps.clock.monotonicMs() + UPSTREAM_PAUSE_MS;
      }
    } catch (err) {
      item.reject(err);
    } finally {
      projectRelease();
      globalRelease();
      this.active--;
      this.notifier.notify();
    }
  }

  private findAdmissible(): Admission | null {
    if (this.deps.globalSem.available <= 0) return null;
    for (const item of this.deps.queue.snapshot()) {
      const globalRelease = this.deps.globalSem.tryAcquire();
      if (!globalRelease) return null;
      const projectRelease = this.deps.projectSem.tryAcquire(item.projectId);
      if (projectRelease) {
        return { item, globalRelease, projectRelease };
      }
      globalRelease();
    }
    return null;
  }

  private diagnoseBlock(): BlockReason {
    if (this.deps.globalSem.available <= 0) return "global-limit";
    if (this.deps.projectSem.activeProjectIds().length >= this.deps.config.limits.maxActiveProjects) {
      return "active-projects";
    }
    return "project-limit";
  }

  private emitBlocked(reason: BlockReason, msUntilRetry?: number): void {
    if (!this.deps.hooks.onBlocked) return;
    this.deps.hooks.onBlocked({
      reason,
      ...(msUntilRetry !== undefined ? { msUntilRetry } : {}),
    });
  }

  private removeFromQueue(id: string): void {
    this.deps.queue.removeById(id);
  }

  private async waitForSignal(maxMs: number): Promise<void> {
    if (maxMs <= 0) return;
    await Promise.race([
      this.deps.clock.sleep(maxMs).catch(() => undefined),
      this.notifier.wait(),
    ]);
  }
}

interface Admission {
  readonly item: PendingTask<TaskResult>;
  readonly globalRelease: () => void;
  readonly projectRelease: () => void;
}

export type { Task };
